File: src/daz_agent_sdk/capabilities/stt.py
from __future__ import annotations

import asyncio
from pathlib import Path
from uuid import UUID

from daz_agent_sdk.logging_ import ConversationLogger
from daz_agent_sdk.types import AgentError, Capability, ErrorKind, ModelInfo, Tier


# ##################################################################
# local stt model info
# placeholder ModelInfo for the local whisper subprocess tool.
_LOCAL_STT_MODEL = ModelInfo(
    provider="local",
    model_id="whisper",
    display_name="Local Whisper STT",
    capabilities=frozenset({Capability.STT}),
    tier=Tier.HIGH,
    supports_streaming=False,
    supports_structured=False,
    supports_conversation=False,
)

# ##################################################################
# default model size
# whisper model size when caller does not specify
_DEFAULT_MODEL_SIZE = "small"


# ##################################################################
# build command
# constructs the whisper subprocess argument list from the supplied
# parameters. audio_path must already be resolved by the caller.
def _build_stt_command(
    audio_path: Path,
    *,
    model_size: str,
    language: str | None,
) -> list[str]:
    cmd = [
        "whisper",
        str(audio_path),
        "--model", model_size,
        "--output_format", "txt",
    ]
    if language is not None:
        cmd.extend(["--language", language])
    return cmd


# ##################################################################
# run subprocess
# runs a subprocess using asyncio.create_subprocess_exec and waits
# for it to complete. returns stdout text. raises AgentError if the
# process exits non-zero or the timeout is exceeded.
async def _run_subprocess(args: list[str], *, timeout: float, label: str) -> str:
    try:
        proc = await asyncio.create_subprocess_exec(
            *args,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        try:
            stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=timeout)
        except asyncio.TimeoutError as exc:
            proc.kill()
            await proc.communicate()
            raise AgentError(
                f"{label} timed out after {timeout}s",
                kind=ErrorKind.TIMEOUT,
            ) from exc

        if proc.returncode != 0:
            stderr_text = stderr.decode(errors="replace").strip()
            raise AgentError(
                f"{label} failed (exit {proc.returncode}): {stderr_text}",
                kind=ErrorKind.INTERNAL,
            )
        return stdout.decode(errors="replace")
    except AgentError:
        raise
    except OSError as exc:
        raise AgentError(
            f"{label} could not be started: {exc}",
            kind=ErrorKind.NOT_AVAILABLE,
        ) from exc


# ##################################################################
# transcribe
# transcribes audio to text using the local whisper subprocess tool.
# model_size controls the whisper model variant (base, small, large-v3-turbo).
# language is optional — whisper auto-detects if not provided.
async def transcribe(
    audio: str | Path,
    *,
    model_size: str = _DEFAULT_MODEL_SIZE,
    language: str | None = None,
    timeout: float = 120.0,
    logger: ConversationLogger | None = None,
    conversation_id: UUID | None = None,
) -> str:
    audio_path = Path(audio)

    if not audio_path.exists():
        raise AgentError(
            f"Audio file does not exist: {audio_path}",
            kind=ErrorKind.INVALID_REQUEST,
        )

    if logger is not None:
        logger.log_event(
            "stt_request",
            audio=str(audio_path),
            model_size=model_size,
            language=language,
        )

    cmd = _build_stt_command(audio_path, model_size=model_size, language=language)
    stdout = await _run_subprocess(cmd, timeout=timeout, label="whisper")
    text = stdout.strip()

    if logger is not None:
        logger.log_event(
            "stt_complete",
            audio=str(audio_path),
            text_length=len(text),
        )

    return text

File: go/capability/stt.go
package capability

import (
	"bytes"
	"context"
	"fmt"
	"os"
	"os/exec"
	"strings"
	"time"

	"github.com/google/uuid"

	agentsdk "github.com/darrenoakey/daz-agent-sdk/go"
)

// defaultModelSize is the whisper model size when the caller does not specify.
const defaultModelSize = "small"

// localSTTModelInfo describes the local Whisper subprocess tool.
var localSTTModelInfo = agentsdk.ModelInfo{
	Provider:     "local",
	ModelID:      "whisper",
	DisplayName:  "Local Whisper STT",
	Capabilities: []agentsdk.Capability{agentsdk.CapabilitySTT},
	Tier:         agentsdk.TierHigh,
}

// TranscribeOpts holds optional parameters for Transcribe.
type TranscribeOpts struct {
	// ModelSize is the whisper model variant (base, small, large-v3-turbo).
	// Empty means "small".
	ModelSize string
	// Language is the expected language code. Empty means auto-detect.
	Language string
	// Timeout is the subprocess timeout. Zero means 120 seconds.
	Timeout time.Duration
	// Logger is an optional conversation logger for event recording.
	Logger *agentsdk.ConversationLogger
	// ConversationID ties the result to a conversation. Zero means generate one.
	ConversationID uuid.UUID
}

// buildSTTCommand constructs the whisper subprocess argument list.
func buildSTTCommand(audioPath string, modelSize string, language string) []string {
	cmd := []string{
		"whisper",
		audioPath,
		"--model", modelSize,
		"--output_format", "txt",
	}
	if language != "" {
		cmd = append(cmd, "--language", language)
	}
	return cmd
}

// Transcribe converts audio to text using the local whisper subprocess tool.
// It returns the transcribed text.
func Transcribe(ctx context.Context, audioPath string, opts TranscribeOpts) (string, error) {
	// Verify audio file exists
	if _, err := os.Stat(audioPath); os.IsNotExist(err) {
		return "", agentsdk.NewAgentError(
			fmt.Sprintf("audio file does not exist: %s", audioPath),
			agentsdk.ErrorInvalidRequest, nil,
		)
	}

	modelSize := opts.ModelSize
	if modelSize == "" {
		modelSize = defaultModelSize
	}

	timeout := opts.Timeout
	if timeout == 0 {
		timeout = 120 * time.Second
	}

	if opts.Logger != nil {
		opts.Logger.LogEvent("stt_request", map[string]any{
			"audio":      audioPath,
			"model_size": modelSize,
			"language":   opts.Language,
		})
	}

	args := buildSTTCommand(audioPath, modelSize, opts.Language)
	stdout, err := runSubprocessWithOutput(ctx, args, timeout, "whisper")
	if err != nil {
		return "", err
	}

	text := strings.TrimSpace(stdout)

	if opts.Logger != nil {
		opts.Logger.LogEvent("stt_complete", map[string]any{
			"audio":       audioPath,
			"text_length": len(text),
		})
	}

	return text, nil
}

// runSubprocessWithOutput executes a command with a timeout, capturing
// both stdout and stderr. Returns stdout text on success, or an AgentError
// on failure or timeout.
func runSubprocessWithOutput(ctx context.Context, args []string, timeout time.Duration, label string) (string, error) {
	cmdCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	cmd := exec.CommandContext(cmdCtx, args[0], args[1:]...)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	if err != nil {
		if cmdCtx.Err() == context.DeadlineExceeded {
			return "", agentsdk.NewAgentError(
				fmt.Sprintf("%s timed out after %s", label, timeout),
				agentsdk.ErrorTimeout, nil,
			)
		}
		if execErr, ok := err.(*exec.Error); ok {
			return "", agentsdk.NewAgentError(
				fmt.Sprintf("%s could not be started: %v", label, execErr),
				agentsdk.ErrorNotAvailable, nil,
			)
		}
		stderrText := stderr.String()
		exitCode := -1
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		}
		return "", agentsdk.NewAgentError(
			fmt.Sprintf("%s failed (exit %d): %s", label, exitCode, stderrText),
			agentsdk.ErrorInternal, nil,
		)
	}
	return stdout.String(), nil
}
